src/Build/Evaluation/Expander.cs

```diff
@@ -2747,6 +2832,111 @@ internal static IEnumerable<KeyValuePair<string, S>> ExpandQuotedExpressionFunct
                             yield return new KeyValuePair<string, S>(null, item.Value);
                         }
                     }
+
+                    s_includeBuilder = includeBuilder;
+                }
+
+                /// <summary>
+                /// This is effectively an "unrolled" while loop, with several manual checks before we enter the full loop,
+                /// so there's an intentional amount of duplication. In the vast majority of cases, we'll only have
+                /// 1-2 matches, and within those we can avoid allocating the vast majority of Regex objects and return
+                /// a cached result.
+                /// </summary>
+                private static OneOrMultipleMetadataMatches GetQuotedExpressionMatches(string quotedExpressionFunction, IElementLocation elementLocation)
+                {
+                    // Start with fast paths to avoid any allocations.
+                    if (TryGetCachedMetadataMatch(quotedExpressionFunction, out string cachedName)
+                        || s_itemSpecModifiers.TryGetValue(quotedExpressionFunction, out cachedName))
+                    {
+                        return new OneOrMultipleMetadataMatches(cachedName);
+                    }
+
+                    Match match = RegularExpressions.ItemMetadataRegex.Match(quotedExpressionFunction);
+
+                    // GroupCollection + Groups are the most expensive source of allocations here, so we want to return
+                    // before ever accessing the property. Simply accessing it will trigger the full collection
+                    // allocation, so we avoid it unless absolutely necessary.
+                    // Unfortunately even .NET Core does not have a struct-based Group enumerator at this point.
+                    if (!match.Success)
+                    {
+                        return new OneOrMultipleMetadataMatches(string.Empty);
+                    }
+                    else if (s_itemSpecModifiers.TryGetValue(match.Value, out cachedName))
```

Comment:
could you elaborate on this case in a comment? it's successful and it's . . . a single match, that isn't exactly one of the well-formed cases but is "close enough"? Is it like whitespace around the exact spelling so `@(Foo->'  %(Identity)  ')`?